text/0001-catalogs.md

```diff
@@ -0,0 +1,341 @@
+
+For example, developers may find + replace `"react": "^18.1.0"` to `"react": "^18.2.0"` across a repository.
+
+- These giant edits lead to churn in `package.json` files and merge conflicts with other commits editing `package.json` files.
+- Repository maintainers have to remember to run this synchronization step periodically, or create a Continuous Integration step to verify the repo is in a good state.
+
+Neither of these error-prone operations are necessary when using a catalog.
+
+### 2. Intention becomes clear
+
+There might be a tight relationship between `foo` and `bar`.
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "^17.0.2"
+  }
+}
+```
+
+```json5
+{
+  "name": "@monorepo/bar",
+  "dependencies": {
+    "react": "^17.0.2"
+  }
+}
+```
+
+A developer working primarily in `@monorepo/bar` may not realize the implied coupling and upgrade `@monorepo/bar` to `react@18` without realizing an edit to `@monorepo/foo` was also required.
+
+The `catalog:` protocol makes it more clear from just reading `package.json` when a dependency is intended to be consistent across the monorepo. Ideally this person would search "catalog package.json` online and find pnpm.io docs.
+
+## Implementation
+
+### Fetching
+
+Catalogs can be shared through standard NPM packages. The `pnpm-catalog-v1` key in `exports` will be used when referencing an external catalog.
+
+```json
+{
+  "name": "@example/frontend-catalog",
+  "exports": {
+    "pnpm-catalog-v1": "./catalog.yaml"
+  }
+}
+```
+
+The schema of a catalog may be extended in the future. In this hypothetical, the `-v1` portion of the exports key enables a package to publish catalogs for multiple versions of pnpm that understand different schema versions.
+
+Only the subset of the version catalog that's used will be fetched and installed.
+
+### Lockfile
+
+Catalogs will be saved to `pnpm-lock.yaml` under a new `catalogs` key. This allows users to more easily review changes to catalogs and pnpm to perform faster up-to-date checks.
+
+```yaml
+lockfileVersion: next
+
+importers:
+
+  packages/foo:
+    dependencies:
+      react:
+        specifier: 'catalog:@example/frontend-catalog'
+        version: ^18.2.0
+
+catalogs:
+
+  /@example/frontend-catalog@0.1.0:
+    resolution: {integrity: sha512...}
+    versions:
+      react:
+        specifier: ^18.2.0
+      react-dom:
+        specifier: ^18.2.0
+      redux:
+        specifier: ^4.2.0
+      react-redux:
+        specifier: ^8.0.0
+
+packages:
+  # ...
+```
+
+### Other Changes
+
+Similar to the [`workspace:` protocol](https://pnpm.io/workspaces#workspace-protocol-workspace), `pnpm publish` will need to replace instances of `catalog:` with valid specifiers before publishing.
```

Comment:
A lot of monorepos I've worked on use `pnpm` to install but use to `npm` to publish because its safer and gets new features like [provenance](https://github.blog/2023-04-19-introducing-npm-package-provenance/).

If catalogs only works with `pnpm` then it could go unused by the monorepos they rely on `npm publish`.

Reply:
but `pnpm publish` runs `npm publish` under the hood.

Reply:
So with this RFC, `pnpm publish` would modify package.json and write it to disk right before running `npm publish`?

Reply:
> So with this RFC, pnpm publish would modify package.json and write it to disk right before running npm publish?

Sort of. The original `package.json` file never gets modified on disk. When `package.json` is read, it's transformed in-memory and passed written to a `.tgz`. The `npm publish` command then sees that `.tgz` instead of the original package directory.

https://github.com/pnpm/pnpm/blob/c099806f1dca360b083f9aeb58848cf6593dcd58/releasing/plugin-commands-publishing/src/pack.ts#L161

It may be worth noting that pnpm wasn't the first package manager implementing the `workspace:` protocol. I believe it was a feature of Yarn v2 originally: https://yarnpkg.com/features/workspaces

The above describes `workspace:`, but I'd imagine `catalog:` to be another step in that existing process.

Reply:
> The original package.json file never gets modified on disk.

Won't that cause problems for other tools that read package.json such such as turborepo?

Or perhaps as long as it exists but the lockfile is updated appropriately? 

https://github.com/pnpm/rfcs/pull/3#issuecomment-1628059128

Reply:
> Or perhaps as long as it package.json exists but the lockfile is updated appropriately then it will work with others?

For catalogs, I believe the answer to this is yes? From the linked comment:

> Dependencies: we resolve pnpm's (and npm's, and yarn's, and berry's) lockfile, not package.json. (This was the work of chris-olszewski and he can provide color commentary if necessary.) We basically don't care what happens here as long as it gets serialized into the lockfile.

Going to respond to the comments in the other thread next.